server/src/delta_cache.rs

```diff
@@ -0,0 +1,286 @@
+                stale: None,
+                impression_data: None,
+                project: None,
+                strategies: None,
+                variants: None,
+                dependencies: None,
+            },
+        ],
+        segments: vec![
+            Segment { id: 1, constraints: vec![] },
+            Segment { id: 2,constraints: vec![] },
+        ],
+    };
+    let max_length = 2;
+    let mut delta_cache = DeltaCache::new(base_event.clone(), max_length);
+
+    let initial_events = vec![
+        DeltaEvent::FeatureUpdated {
+            event_id: 2,
+            feature: ClientFeature {
+                name: "my-feature-flag".to_string(),
+                feature_type: None,
+                description: None,
+                created_at: None,
+                last_seen_at: None,
+                enabled: true,
+                stale: None,
+                impression_data: None,
+                project: None,
+                strategies: None,
+                variants: None,
+                dependencies: None,
+            },
+        },
+    ];
+    delta_cache.add_events(initial_events);
+
+    let added_events = vec![
+        DeltaEvent::FeatureUpdated {
+            event_id: 3,
+            feature: ClientFeature {
+                name: "another-feature-flag".to_string(),
+                feature_type: None,
+                description: None,
+                created_at: None,
+                last_seen_at: None,
+                enabled: true,
+                stale: None,
+                impression_data: None,
+                project: None,
+                strategies: None,
+                variants: None,
+                dependencies: None,
+            },
+        },
+        DeltaEvent::FeatureRemoved {
+            event_id: 4,
+            feature_name: "test-flag".to_string(),
+        },
+        DeltaEvent::SegmentUpdated {
+            event_id: 5,
+            segment: Segment { id: 1,  constraints: vec![] },
+        },
+        DeltaEvent::SegmentRemoved {
+            event_id: 6,
+            segment_id: 2,
+        },
+        DeltaEvent::SegmentUpdated {
+            event_id: 7,
+            segment: Segment { id: 3, constraints: vec![] },
+        },
+    ];
+    delta_cache.add_events(added_events.clone());
+
+    let events: Vec<_> = delta_cache.get_events().iter().cloned().collect();
+    assert_eq!(events.len(), max_length);
+    assert_eq!(events, added_events[added_events.len() - max_length..]);
+
+    let hydration_event = delta_cache.get_hydration_event();
+    assert_eq!(hydration_event.features.len(), 2);
+    assert_eq!(hydration_event.event_id, 7);
+    assert!(hydration_event.features.iter().any(|f| f.name == "my-feature-flag"));
+    assert!(hydration_event.features.iter().any(|f| f.name == "another-feature-flag"));
+    assert!(hydration_event.segments.iter().any(|s| s.id == 1));
+}
+
+#[test]
+fn test_prevent_mutation_of_previous_feature_updated_events() {
```

Comment:
For my understanding, you're testing that adding two FeatureUpdated events updating the same event keeps both of them?

Reply:
Yes